Python/specialize.c

```diff
@@ -1539,3 +1551,48 @@ _Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
     STAT_INC(BINARY_OP, specialization_success);
     adaptive->counter = initial_counter_value();
 }
+
+void
+_Py_Specialize_CompareOp(PyObject *lhs, PyObject *rhs,
+                         _Py_CODEUNIT *instr, SpecializedCacheEntry *cache)
+{
+    _PyAdaptiveEntry *adaptive = &cache->adaptive;
+    if (Py_TYPE(lhs) != Py_TYPE(rhs)) {
+        SPECIALIZATION_FAIL(COMPARE_OP, SPEC_FAIL_DIFFERENT_TYPES);
+        goto failure;
+    }
+    else if (PyFloat_CheckExact(lhs)) {
+        *instr = _Py_MAKECODEUNIT(COMPARE_OP_FLOAT, _Py_OPARG(*instr));
+        goto success;
+    }
+    else if (PyLong_CheckExact(lhs)) {
+        *instr = _Py_MAKECODEUNIT(COMPARE_OP_INT, _Py_OPARG(*instr));
+        goto success;
+    }
+    else if (PyUnicode_CheckExact(lhs)) {
+        int op = adaptive->original_oparg;
+        if (op != Py_EQ && op != Py_NE) {
+            SPECIALIZATION_FAIL(COMPARE_OP, SPEC_FAIL_STRING_COMPARE);
+            goto failure;
+        }
+        else if (!PyUnicode_IS_READY(lhs) || !PyUnicode_IS_READY(rhs)) {
```

Comment:
There is no point in this check. The instruction needs to check for readiness regardless.

Reply:
If a bytecode is going to fail the readiness check during specialization, isn't it likely to fail the readiness check afterwards? My worry was that without the check, the following code would needlessly re-optimize and de-optimize repeatedly:

```
for s in unready_strings_from_c_extension():
    if s == "Python":
        break
```